Fix Ctrl-C segfaults in sage.libs.gap

orlitzky opened the pull request:
Replace cysignals' `sig_on()` and `sig_off()` with custom `gap_sig_on()` and `gap_sig_off()` wrappers that use GAP's own `SIGINT` handling. This fixes an uncommon, but ultimately reproducible segfault when Ctrl-C is used to interrupt a GAP computation.

In concert with https://github.com/sagemath/sage/pull/40594 this has allowed `sage/libs/gap/element.pyx` to pass many thousands of test iterations.

### Fixes

* https://github.com/sagemath/sage/issues/40598

### Dependencies

* https://github.com/sagemath/sage/pull/40594

user202729 committed:
Fix segmentation fault in libgap function call

user202729 committed:
Code modification to avoid calling GapElement.__dealloc__ too early

orlitzky committed:
src/sage/libs/gap/meson.build: remove cysignals dependency

We are no longer using cysignals for this; instead we have our own
gap_sig_on() and gap_sig_off() functions.

orlitzky committed:
src/sage/libs/gap/gap_includes.pxd: add InterruptExecStat

We plan to install InterruptExecStat (gap/stats.h) as the SIGINT and
SIGALRM handler while GAP code is running, since that seems to work
better than mixing cysignals with GAP_Enter/GAP_Leave has.

orlitzky committed:
src/sage/doctest/util.py: handle Ctrl-C from GAP

We've found that GAP's own SIGINT handler is less crashy than if we
mix GAP_Enter/GAP_Leave with cysignals' sig_on and sig_off. We've
installed that same handler for SIGALRM, but the code that catches it
can't tell the difference and converts them both in to
KeyboardInterrupt. To retain the ability to doctest this stuff, we have
to catch those KeyboardInterrupts and poke at them to see if they arose
from GAP.

orlitzky committed:
src/sage/libs/gap/util.pyx: add gap_sig_on / gap_sig_off

Add two new pre/post-GAP functions to enable/disable GAP's own SIGINT
(Ctrl-C) handler. These avoid the setjmp/longjmp issues we've
encountered with the sig_on() and sig_off() from cysignals.

orlitzky committed:
src/sage/libs/gap/util.pyx: replace sig_on / sig_off

Replace the sig_on() and sig_off() wrappers from cysignals with the
new custom gap_sig_on() and gap_sig_off(). We've lost the sig_on() and
sig_off() around GAP_initialize() because I don't think we can trust
GAP's own signal handler until after we've initialized GAP.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: update signal handling docs

The way we handle signals in libgap is now unusual, so a few
paragraphs have been added to the libgap module docstring explaining
how and why it is unusual.

orlitzky committed:
src/sage/libs/gap/element.pyx: replace sig_on / sig_off

Replace the sig_on() and sig_off() wrappers from cysignals with the
new custom gap_sig_on() and gap_sig_off(). This fixes the segfault
that occurs after repeated testing of _pow_.

orlitzky committed:
src/sage/libs/gap/gap_includes.pxd: drop sig_GAP_Enter()

This function used cysignals for sig_error(), and we have opted not to
mix cysignals code with libgap code. It has in any case been replaced
by plain GAP_Enter calls.

user202729 commented on src/sage/libs/gap/libgap.pyx:
```diff
@@ -160,11 +160,43 @@ using the recursive expansion of the
+   used to wrap code that may take a long time, and, as a result, may
+   need to be interrupted with Ctrl-C. Those functions are implemented
+   with setjmp and longjmp however, and cause hard-to-diagnose issues
+   arising from their interaction with ``GAP_enter()`` and
```
```suggestion
   arising from their interaction with ``GAP_Enter()`` and
```

That said, the description is actually misleading. It's not the interference with `GAP_Enter` and `GAP_Leave` that causes the issue, it's `longjmp` from a signal handler itself. (there are other deadlocks and CI timeouts elsewhere too, and they does not have interference with another `setjmp`/`longjmp`.)

I suggest something like

Typically, cysignals' sig_on() and sig_off() functions are used to wrap code that may take a long time, and as a result, may need to be interrupted with Ctrl-C. However, it is possible that interrupting a function execution at an arbitrary location will lead to inconsistent state. Internally, GAP provides a mechanism using ``InterruptExecStat``, which sets a flag that tells GAP to gracefully exit with an error as early as possible. We make use of this internal mechanism to prevent segmentation fault when GAP functions are interrupted.

orlitzky replied:
Ok, you're right, I can't explain exactly what is causing the segfault, so the less we say the better. I've used your explanation (which may need to be updated later).

user202729 commented on src/sage/libs/gap/util.pyx:
```diff
@@ -283,6 +299,23 @@ cdef initialize():
+    global gap_sigint_sa
+    global sage_sigint_sa
+    global sage_sigalrm_sa
+    sigaction(SIGINT, &gap_sigint_sa, &sage_sigint_sa)
```
is it possible that before setting the signal handler it is `SIG_DFL` or `SIG_IGN`? If so, does this correctly restore the signal handler?

orlitzky replied:
It looks like it, if I write a short C program that uses `signal()` to set the handler to either `SIG_DFL` or `SIG_IGN` and then change the handler with `sigaction()`, the old action info is saved (e.g. `sa_handler == SIG_DFL` is true).

user202729 replied:
We may also want to ask whether this behavior is _intended_ i.e. if the user go out of their way to set `SIG_IGN`, maybe we could keep ignoring it (though no strong opinion either way)

orlitzky replied:
I think we should wait for at least one person to complain about this before we jump through any hoops. The same change would need to be made in several places.

user202729 commented on src/sage/libs/gap/libgap.pyx:
```diff
@@ -160,11 +160,43 @@ using the recursive expansion of the
+
+   The approach we are using instead is to install GAP's own
+   ``SIGINT`` handler (to catch Ctrl-C) before executing any
+   long-running GAP code, and then to later uninstall it when the GAP
```
maybe change "uninstall it" to "restore the previous signal handler".

orlitzky replied:
done

user202729 commented on src/sage/libs/gap/element.pyx:
```diff
@@ -1104,12 +1143,18 @@ cdef class GapElement(RingElement):
+                # Ctrl-C
+                raise KeyboardInterrupt from e
+            else:
+                raise
```
this is some really bad duplication of code. I suggest 

* fold `GAP_Enter` and `GAP_Leave` into `gap_sig_on` and `gap_sig_off` respectively
* instead of setting the interrupt handler directly to `InterruptExecStat`, make custom wrapper function (or hook into cysignals), the custom wrapper function sets a global variable that determine which kind of signal is being raised, then call `InterruptExecStat`
* the global variable above need to be reset at appropriate places
* modify `error_handler` to, instead of indiscriminately raise `GAPError`, raise the correct exception by reading the global variable above

orlitzky replied:
Yeah it's pretty bad but I was trying to keep it as simple as possible for the first iteration. First get it to not crash, then make it pretty.

I don't think we can use a global for the signal type because signals are asynchronous. There's no reason why a `SIGINT` can't be triggered while handling a `SIGALRM`, or vice-versa. It's not very likely, but if there's a lesson here it's that we shouldn't count on unlikely things not happening.

We may be able to combine `GAP_Enter` and `gap_sig_on()`, though.

user202729 replied:
> I don't think we can use a global for the signal type because signals are asynchronous. There's no reason why a `SIGINT` can't be triggered while handling a `SIGALRM`, or vice-versa. It's not very likely, but if there's a lesson here it's that we shouldn't count on unlikely things not happening.

you're right but… I think currently the signal handler is calling `PyErr_Restore()` or something anyway, and that one isn't reentrant either.

Searching up a bit https://stackoverflow.com/a/3127697/, that's indeed a possibility. But doesn't the same thing happen for cysignals as well?

orlitzky replied:
> We may be able to combine `GAP_Enter` and `gap_sig_on()`, though.

Even this immediately leads to a segault.

user202729 commented on src/sage/doctest/util.py:
```diff
@@ -886,6 +886,19 @@ def ensure_interruptible_after(seconds: float, max_wait_after_interrupt: float =
+            e.__traceback__ = None
+            alarm_raised = True
+        else:
+            raise
```
if the logic below (raise AlarmInterrupt instead of KeyboardInterrupt) are correctly implemented, this change wouldn't be necessary.

orlitzky replied:
I've further simplified this down to a one-line change, using the fact that `AlarmInterrupt` is a subclass of `KeyboardInterrupt`. Instead of catching `AlarmInterrupt`, we can catch `KeyboardInterrupt`, and then check,

* is it an `AlarmInterrupt`?
* does it have "user interrupt" as the message?

I prefer this to hacking `AlarmInterrupt` into the GAP code at this point because `sage.libs.gap` is otherwise free of cysignals. To raise the `AlarmInterrupt` from `sage.libs.gap` we'd have to,

1. Add cysignals back as a dep in `meson.build`
2. Import `AlarmInterrupt`
3. Add the global `last_signal` variable
4. Set `last_signal` whenever a handler is called
5. Check `last_signal` when raising a `GAPError` and convert it to the right thing

Since this is all for the benefit of one doctest method, it just seems easier to add the one line in that doctest method?

orlitzky committed:
src/sage/libs/gap/libgap.pyx: rewrite signal handling explanation

orlitzky committed:
src/sage/libs/gap/util.pyx: do KeyboardInterrupt conversion earlier

If we're going to raise a GAPError that resulted from a SIGINT or
SIGALRM, we might as well turn it in to a KeyboardInterrupt right
away.

orlitzky committed:
src/sage/doctest/util.py: update GAP workaround

We are no longer retaining the __cause__ of KeyboardInterrupts
that arise from GAP, but the error message is still there, so
we can use that to filter these out.

orlitzky committed:
src/sage/libs/gap: remove redundant KeyboardInterrupt hacks

We are now converting "GAPError: Error, user interrupt" into a
KeyboardInterrupt in our libgap error handler, so we don't have
to do it at every call site.

user202729 commented on src/sage/libs/gap/util.pyx:
```diff
@@ -258,11 +264,21 @@ cdef initialize():
+    # own SIGINT handler (but without the double-Ctrl-C behavior), and
+    # is less crashy than when we mix cysignals with GAP code.
+    global gap_sigint_sa
+    gap_sigint_sa.sa_handler = <void (*)(int) noexcept>InterruptExecStat
```
no, casting function from one pointer type to another is directly undefined behavior. Define a wrapper function in C (not in Cython!) that wraps this function.

Then you can execute arbitrary logic in that function, including distinguishing sigint and sigalrm.

orlitzky replied:
You can do it so long as the types are "compatible," but I don't have any faith in my ability to interpret that part of the C standard, so I've wrapped it.

user202729 commented on src/sage/libs/gap/gap_includes.pxd:
```diff
@@ -40,7 +37,6 @@ cdef extern from "gap/libgap-api.h" nogil:
     cdef void GAP_EnterStack()
     cdef void GAP_LeaveStack()
     cdef int GAP_Enter() except 0
```
actually looking at it again, the `except 0` guarantees that `GAP_Enter` will raise the error (provided it's already set). So the logic before this change is reasonably safe.

orlitzky replied:
FFS, good catch

orlitzky committed:
src/sage/libs/gap/util.pyx: type-safety for InterruptExecStat()

Wrap InterruptExecStat() in a function that takes an int so that the
associated function pointer is what sigaction_t.sa_handler expects

orlitzky committed:
src/sage/doctest/util.py: further simplify GAP workaround

Since AlarmInterrupt is a subclass of KeyboardInterrupt, we can catch
them both with "except KeyboardInterrupt" and then filter out the ones
we want. This reduces a small amount of code duplication.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: explain error handling

The way errors are handled in this interface is tricky. It is now
explained to the best of my ability in the module's docstring, and the
(related) paragraphs about signal handling have been updated as well.

orlitzky committed:
src/sage/libs/gap/util.pyx: delete old commented debug statement

orlitzky committed:
src/sage/libs/gap: use public API for is_string()

We have GAP_IsString() in the public API now, it is preferable to
using the internal IS_STRING.

orlitzky committed:
src/sage/libs/gap: be more careful inside make_any_gap_element()

This function performs a bunch of heuristics to decide what type of
GAP element to make. In doing so, it jumps between libgap (C) code
sage (python) code, all while inside of a GAP_Enter / GAP_Leave
pair. Moreover, some of the functions it calls are wrapped in their
own GAP_Enter / GAP_Leave pairs.

This commit reorganizes things so that the libgap code, wrapped in
GAP_Enter and GAP_Leave, comes first. The libgap code decides what
type of GAP element we have. It is then followed by the python case
statement that decides what type of Sage element we'll construct.

We've also dropped the internal IsStringConv() function in favor of
the public GAP_IsString(). The latter does not convert a list of
characters to an efficient string representation, but modulo some
doctests that have been updated, that should not be a problem: if you
don't want a list of characters, don't pass in a list of characters.

orlitzky committed:
src/sage/libs/gap/element.pyx: missing GAP_Enter / GAP_Leave

The is_list() and is_record() methods call libgap, and need to be
sandwiched between GAP_Enter and GAP_Leave.

orlitzky committed:
src/sage/libs/gap/element.pyx: drop superclass list -> sage conversion

There's already a sage() method in the GapElement_List subclass. If
you have a list but you wound up with a generic GapElement instead of
a GapElement_List, you probably used the API wrong.

user202729 commented on src/sage/libs/gap/libgap.pyx:
```diff
@@ -160,11 +160,108 @@ using the recursive expansion of the
+
+       if (GAP_Enter()):
+          # further calls to libgap
+          GAP_Leave()
```
I suggest something like::

> even though the following is correct in C:
> 
> ```
> int ok = GAP_Enter()
> if (ok) {
>     # further call to libgap
> }
> GAP_Leave()
> ```
> 
> the following similar-looking code is incorrect in Cython:
> 
> ```
> cdef int ok
> ok = GAP_Enter()
> if ok:
>     # further call to libgap
> GAP_Leave()
> ```
> 
> ... further explanation ...

or you may want to make a minimal change and un-indent the `GAP_Leave` by 4 spaces.

Then later also document what's the correct way to call `sig_gap_enter` etc.
It may be useful to also mention that unlike `sig_on()`, `GAP_Enter()` must be inside the `try:`.

orlitzky replied:
I deleted this example. I was talking to myself. Nobody else is going to want to read what **not** to do, and having a bad example in there risks someone copy/pasting it without reading the surrounding paragraph.

user202729 commented on src/sage/libs/gap/libgap.pyx:
```diff
@@ -160,11 +160,108 @@ using the recursive expansion of the
+           GAP_Enter()
+           libgap.Sum(1,2,3,4)
+       finally:
+           GAP_Leave()
```
The point is we only use `GAP_Enter/Leave` to wrap around calls to C functions, `libgap.Sum` already use it internally.

So this example is actually confusing, because **the user** shouldn't be writing code like this.

same thing to be said with the explanation above:

> All GAP library usage

here "library" may refer to the C API by GAP or it may be misunderstood as the `libgap` Python API by Sage.

orlitzky replied:
You are right, I've come up with a better example, and have clarified that we mean the GAP C library.

user202729 commented on src/sage/libs/gap/element.pyx:
```diff
@@ -296,53 +322,82 @@ cdef GapElement make_any_gap_element(parent, Obj obj):
+
+    # Plan B: Use the "private" API to determine the type of ``obj``.
+    # Since TNUM_OBJ does not invoke the public libgap API, it is not
+    # wrapped in GAP_Enter / GAP_Leave.
```
this explanation is dubious. If public API function A calls private API function B, and function B does something with the garbage collector, surely you need GAP_Enter/GAP_Leave too?

That said, a good explanation would be "the current implementation of these functions does not mutate any state or invoke the garbage collector, so GAP_Enter/GAP_Leave is unnecessary". That assumption, of course, is dangerous, because what if it does mutate some state/potentially invoke the garbage collector later.

Anyway, if you agree that we can use this assumption, we could just not have GAP_Enter/Leave here at all, and considerably shorten the code (e.g. you can delete the enum above)

dimpase replied:
Does (lib)GAP has a way to ensure that GC is not called during a call of a GAP function?
I don't know - perhaps this ought to be requested upstream...

orlitzky replied:
> Does (lib)GAP has a way to ensure that GC is not called during a call of a GAP function? I don't know - perhaps this ought to be requested upstream...

When we construct a GAP element, the `_initialize` method creates a global reference to it, and then later...

```python
cdef void gasman_callback() noexcept with gil:                                       
    """                                                                              
    Callback before each GAP garbage collection                                      
    """                                                                              
    global owned_objects_refcount                                                    
    for obj in owned_objects_refcount:                                               
        GAP_MarkBag((<ObjWrapper>obj).value)
```

Supposedly that prevents the referenced objects from being nuked:

```C
// Manual management of the GAP garbage collector: for cases where you want          
// a GAP object to be long-lived beyond the context of the stack frame where         
// it was created, it is necessary to call GAP_MarkBag on the object when            
// the garbage collector is run by the markBagsCallback function passed to           
// GAP_Initialize.                                                                   
void GAP_MarkBag(Obj obj);
```

orlitzky replied:
> this explanation is dubious. If public API function A calls private API function B, and function B does something with the garbage collector, surely you need GAP_Enter/GAP_Leave too?

It's more dubious than you think. Unless you read the source, the answer is really "who knows." The `GAP_Enter` docs say call it for libgap-api functions, and `TNUM_OBJ` isn't a libgap-api function; there's my reasoning :)

Historically I think `TNUM_OBJ` and most of the other "internal" functions came before `GAP_Enter` even existed. @dimpase was present when the library interface was born and may remember the details.

In any case, I hate these internal calls, so I've replaced `TNUM_OBJ` with GAP's own `IsFoo()` calls.

orlitzky replied:
> In any case, I hate these internal calls, so I've replaced `TNUM_OBJ` with GAP's own `IsFoo()` calls.

I had to revert this, the speed penalty was too severe. I've updated the comment.

user202729 commented on src/sage/libs/gap/element.pyx:
```diff
@@ -276,16 +287,31 @@ cdef GapElement make_any_gap_element(parent, Obj obj):
 
-        sage: libgap.eval("['a', 'b', 'c']")   # gap strings are also lists of chars
+    Strings in GAP are lists of characters and vice-versa, but that
+    isn't an efficient representation. Keep this in mind if you call a
```
I think you want to say "that isn't an efficient representation **in Python**". surely it should be an efficient representation in GAP (otherwise they wouldn't have done that) right?

edit: okay I think I understood, you want to say that

* to the user, there's no difference between string and list of characters in GAP.
* internally, there are two possible representations of such an object: either a contiguous array of characters, or (inefficiently) a list of characters.

since GAP users don't see this, we need not tell Sage users too much about it either. Best leave to `TESTS::`.

orlitzky replied:
List of characters is always going to be inefficient, but in this case I was referring to the distinction that GAP makes. GAP users do need to be aware of it and there are public functions that convert the list representation into an efficient one.

It's a little confusing because gap shows you "foo" regardless of whether "foo" is efficiently represented, or is represented as `['f','o','o']`. Sage, meanwhile, will show you `['f','o','o']` if that's what you've got.

The `make_any_gap_element()` function was doing the conversion unconditionally, making it impossible to construct a list of characters, and wasting time in the common case where you've passed in a python string and not a python list-of-chars. I undid that; at which point the GAP representation details become important to sage devs.

But that's all irrelevant if we don't want the backwards-incompatible change. I don't care that much about it so I'll put the conversion back for now and leave this test alone.

orlitzky replied:
The backwards-incompatible change has been reverted, so this doctest is back to what it was.

user202729 commented on src/sage/libs/gap/element.pyx:
```diff
@@ -276,16 +287,31 @@ cdef GapElement make_any_gap_element(parent, Obj obj):
+        ['c', 'c']]
         sage: t[1]
         "ab"
         sage: t[1].sage()
```
This is backwards incompatible change. Always return a string in this case.

orlitzky replied:
I reimplemented the list-of-char -> efficient-string conversion, this time using GAP function calls rather than internal library functions. The old behavior is back.

orlitzky committed:
src/sage/libs/gap/element.pyx: revert list-of-char changes

An earlier commit dropped the implicit conversion of (non-empty)
lists-of-char to efficient strings in GAP. Here we reinstate the old
behavior and the associated doctest because it was backwards-
incompatible.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: more realistic "real" example

A real example for GAP_Enter / GAP_Leave should only have libgap C
code sandwiched between the two.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: clarify "GAP library usage"

We specificaly mean the GAP C library, declared in libgap-api.h.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: fix module docstring indentation

Somehow we wound up with a three-space indentation. Make it four.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: more module docstring tweaks

Include an example of "correct" interruptible libgap usage, and delete
an example of what not to do. Bad examples can be dangerous for people
who are skimming the documentation, and are not that interesting
anyway (wrong code is easy to write).

orlitzky committed:
src/sage/libs/gap/util.pyx: add missing noexcept annotation

orlitzky committed:
src/sage/libs/gap/element.pyx: more tolerance in Ctrl-C tests

The check for _pow_ being interruptible by Ctrl-C is subject to some
variation due to CPU usage, scheduling, etc. Here we increase the
tolerance to 5s, which should be more than enough time to interrupt
it, and also increase the size of the problem so that it runs for
at least 5s even on fast machines.

user202729 commented on src/sage/libs/gap/libgap.pyx:
```diff
@@ -160,112 +160,117 @@ using the recursive expansion of the
+    example::
+
+        cpdef void crash_and_burn() except *:
+            cdef GapElement x = <GapElement>libgap({'a': 1, 'b': 2})
```
is this dangerous? The Python wrapper `libgap(…)` is destroyed while the `GapElement` lives on.

orlitzky replied:
Hm, possibly. I updated it to `x = libgap({'a': 1, 'b': 2})` so that it gets a local python reference. But now I'm not so sure that the `xlen = GAP_LenList((<GapElement>x).value)` later on is guaranteed to prevent `x` from being garbage-collected.

FWIW the original pattern appears in `make_gap_list()` and `make_gap_matrix()`. If it's buggy, we should eventually look at those too.

orlitzky committed:
src/sage/libs/gap: special case for bool in make_any_gap_element()

We can check for a GAP boolean using only libgap-api.h by testing for
one of the three objects: GAP_True, GAP_False, GAP_Fail.

orlitzky committed:
src/sage/libs/gap/element.pyx: update make_any_gap_element() comment

We should mention that TNUM_OBJ is safe to use outside of GAP_Enter /
GAP_Leave, and that we are only using it because the alternative is
too slow.

orlitzky committed:
src/sage/libs/gap/element.pyx: faster list length check

We can use the official C API for the length of a list in
make_any_gap_element() without wasting any time.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: slightly safer example

In the GAP_Enter / GAP_Leave example, it's probably better to store
the GapElement in a python variable (rather than a C variable), just
in case python might decide to garbage collect it otherwise.

orlitzky committed:
src/sage/rings/complex_arb.pyx: add some Ctrl-C tolerance to a doctest

We have one doctest in this file that is perpetually failing on the CI
because a CBF integral takes too long to interrupt. The computation is
extremely long-running, though, and "too long" is on the order of
deciseconds: everything is fine. This commit increases the wait
tolerance from 0.2s to 5s. That should be long enough to avoid all
random CPU scheduling issues, and is still much shorter than the time
it would take for the integral to complete.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: unindent the "Using the .. library" docs

These never should have been indented in the first place; the whole
block is being rendered as a quote.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: unindent the AUTHORS block

This is being rendered as a quote.

orlitzky committed:
src/sage/libs/gap/libgap.pyx: backticks for ``libgap-api.h``
